tests/std/tests/P2136R3_invoke_r/test.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <cassert>
+#include <functional>
+#include <type_traits>
+
+using namespace std;
+
+// TRANSITION, DevCom-1457457
+namespace detail {
+    static constexpr bool permissive() {
+        return false;
+    }
+
+    template <class>
+    struct DependentBase {
+        static constexpr bool permissive() {
+            return true;
+        }
+    };
+
+    template <class T>
+    struct Derived : DependentBase<T> {
+        static constexpr bool test() {
+            return permissive();
+        }
+    };
+} // namespace detail
+constexpr bool is_permissive = detail::Derived<int>::test();
+
+constexpr int square(int n) {
+    return n * n;
+}
+constexpr int square_noexcept(int n) noexcept {
+    return n * n;
+}
+
+struct Thing {
+    int n = 0;
+    constexpr int& foo() {
+        return n;
+    }
+};
+
+constexpr bool is_rvalue(const int&) {
+    return false;
+}
+constexpr bool is_rvalue(int&&) {
+    return true;
+}
+
+constexpr bool test_invoke_r() {
+    auto v1 = invoke_r<long int>(square, 3);
+    assert(v1 == 9L);
+    static_assert(is_same_v<decltype(v1), long int>);
+
+    auto v2 = invoke_r<double>([]() -> int { return 5; });
+    assert(v2 == 5);
+    static_assert(is_same_v<decltype(v2), double>);
+    static_assert(is_void_v<decltype(invoke_r<void>(square, 1))>);
+
+    // TRANSITION, DevCom-1457457
+    static_assert(noexcept(invoke_r<int>(square, 3)) == is_permissive, "invoke_r<int>(square, 3) is noexcept");
+    static_assert(noexcept(invoke(square, 3)) == is_permissive, "invoke(square, 3) is noexcept");
+#ifdef __cpp_noexcept_function_type
+    static_assert(noexcept(invoke_r<int>(square_noexcept, 3)), "invoke_r<int>(square_noexcept, 3) isn't noexcept");

```

Comment:
This test coverage is great, but I think we need to cover one additional thing. The Standard (and the implementation here) uses `is_nothrow_invocable_r_v` to examine whether the return type conversion throws. It would be possible for a mistaken implementation (or one damaged during refactoring) to omit the `_r_` part. I think this should be as simple as verifying that plain `invoke` of a `const char * whatever() noexcept` is `noexcept`, but `invoke_r<string>` is not `noexcept` (because the `string(const char *)` constructor can throw exceptions).

Reply:
Should it be testing that they are false when `__cpp_noexcept_function_type` isn't defined?

Reply:
That seems reasonable - it is a (grudgingly) supported mode.